models/gif_psc_exp.cpp

```diff
@@ -0,0 +1,486 @@
+    if ( tau_sfa_[ i ] <= 0 )
+      throw BadProperty( "All time constants must be strictly positive." );
+
+  for ( size_t i = 0; i < tau_stc_.size(); i++ )
+    if ( tau_stc_[ i ] <= 0 )
+      throw BadProperty( "All time constants must be strictly positive." );
+
+
+  if ( tau_ex_ <= 0 || tau_in_ <= 0 )
+    throw BadProperty( "Synapse time constants must be strictly positive." );
+}
+
+void
+nest::gif_psc_exp::State_::get( DictionaryDatum& d, const Parameters_& p ) const
+{
+  def< double >( d, names::V_m, y3_ );    // Membrane potential
+  def< double >( d, names::E_sfa, sfa_ ); // Adaptive threshold potential
+  def< double >( d, names::stc, stc_ );   // Spike triggered current
+}
+
+void
+nest::gif_psc_exp::State_::set( const DictionaryDatum& d, const Parameters_& p )
+{
+  updateValue< double >( d, names::V_m, y3_ );
+  sfa_stc_initialized_ =
+    false; // vectors of the state should be initialized with new parameter set.
+}
+
+nest::gif_psc_exp::Buffers_::Buffers_( gif_psc_exp& n )
+  : logger_( n )
+{
+}
+
+nest::gif_psc_exp::Buffers_::Buffers_( const Buffers_&, gif_psc_exp& n )
+  : logger_( n )
+{
+}
+
+/* ----------------------------------------------------------------
+ * Default and copy constructor for node
+ * ---------------------------------------------------------------- */
+
+nest::gif_psc_exp::gif_psc_exp()
+  : Archiving_Node()
+  , P_()
+  , S_()
+  , B_( *this )
+{
+  recordablesMap_.create();
+}
+
+nest::gif_psc_exp::gif_psc_exp( const gif_psc_exp& n )
+  : Archiving_Node( n )
+  , P_( n.P_ )
+  , S_( n.S_ )
+  , B_( n.B_, *this )
+{
+}
+
+/* ----------------------------------------------------------------
+ * Node initialization functions
+ * ---------------------------------------------------------------- */
+
+void
+nest::gif_psc_exp::init_state_( const Node& proto )
+{
+  const gif_psc_exp& pr = downcast< gif_psc_exp >( proto );
+  S_ = pr.S_;
+}
+
+void
+nest::gif_psc_exp::init_buffers_()
+{
+  B_.spikes_ex_.clear(); // includes resize
+  B_.spikes_in_.clear(); // includes resize
+  B_.currents_.clear();  //!< includes resize
+  B_.logger_.reset();    //!< includes resize
+  Archiving_Node::clear_history();
+}
+
+void
+nest::gif_psc_exp::calibrate()
+{
+
+  B_.logger_.init();
+
+  double_t h = Time::get_resolution().get_ms();
+  V_.rng_ = kernel().rng_manager.get_rng( get_thread() );
+
+  V_.P11ex_ = std::exp( -h / P_.tau_ex_ );
+  V_.P11in_ = std::exp( -h / P_.tau_in_ );
+
+  double_t tau_m = P_.c_m_ / P_.g_L_;
+
+  // these are determined according to a numeric stability criterion
+  V_.P21ex_ = propagator_32( P_.tau_ex_, tau_m, P_.c_m_, h );
+  V_.P21in_ = propagator_32( P_.tau_in_, tau_m, P_.c_m_, h );
+
+
+  V_.P33_ = std::exp( -h / tau_m );
+  V_.P30_ = 1 / P_.c_m_ * ( 1 - V_.P33_ ) * tau_m;
+  V_.P31_ = ( 1 - V_.P33_ );
+
+  V_.RefractoryCounts_ = Time( Time::ms( P_.t_ref_ ) ).get_steps();
+  assert( V_.RefractoryCounts_
+    >= 0 ); // since t_ref_ >= 0, this can only fail in error
```

Comment:
For readbility, I would put the comment on a separate line before the assertion and the assertion on a single line.